sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -405,6 +409,246 @@ protected EventProcessor()
                                                                            string eventHubName,
                                                                            TimeSpan ownershipExpiration) =>
             new PartitionLoadBalancer(storageManager, identifier, consumerGroup, fullyQualifiedNamespace, eventHubName, ownershipExpiration);
+        /// <summary>
+        ///   Creates the infrastructure for tracking the processing of a partition and begins the begins processing the
+        ///   partition in the background until cancellation is requested.
+        /// </summary>
+        ///
+        /// <param name="partition">The Event Hub partition whose processing should be started.</param>
+        /// <param name="startingPosition">The position within the event stream that processing should begin.</param>
+        /// <param name="cancellationSource">A <see cref="CancellationTokenSource"/> instance to signal the request to cancel the operation.</param>
+        ///
+        /// <returns>The <see cref="PartitionProcessor" /> encapsulating the processing task, its cancellation token, and associated state.</returns>
+        ///
+        /// <remarks>
+        ///   This method makes liberal use of class methods and state in addition to the received parameters.
+        /// </remarks>
+        ///
+        internal virtual PartitionProcessor CreatePartitionProcessor(TPartition partition,
+                                                                     EventPosition startingPosition,
+                                                                     CancellationTokenSource cancellationSource)
+        {
+            cancellationSource.Token.ThrowIfCancellationRequested<TaskCanceledException>();
+            var consumer = default(TransportConsumer);
+
+            // If the tracking of the last enqueued event properties was requested, then read the
+            // properties from the active consumer, which can change during processing in the event of
+            // error scenarios.
+
+            LastEnqueuedEventProperties readLastEnquedEventInformation()
+            {
+                if (!Options.TrackLastEnqueuedEventProperties)
+                {
+                    throw new InvalidOperationException(Resources.TrackLastEnqueuedEventPropertiesNotSet);
+                }
+
+                if (consumer == null)
```

Comment:
This should not occur under normal circumstances; the guard exists only to prevent a race condition that is unlikely, but possible, when partition processing is being stopped.

Reply:
Decided this warranted a code comment.

Reply:
I didn't intend to null it out, but certainly open to it.   I was guarding against the scenario where creation somehow failed due to an issue in connection or its linked AmqpScope, the local function could see a null value if called before the partition was stolen or noticed to be in a failed state.  Granted, very unlikely, but possible.

I bigger issue is that I missed an important condition in checking for close in addition.   The race that I'm thinking of here is the one where the background task is shutting down, which closes the consumer at the same time that `EventProcessor<T>.ReadLastEnqueuedEventProperties` gets called.   That could also be the state in the event that the consumer was considered invalid and closed down the processing loop.